// nx/src/lock-file/pnpm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prunePnpmLockFile = exports.transitiveDependencyPnpmLookup = exports.stringifyPnpmLockFile = exports.parsePnpmLockFile = void 0;
const tslib_1 = require("tslib");
const js_yaml_1 = require("@zkochan/js-yaml");
const utils_1 = require("./utils");
const semver_1 = require("semver");
const LOCKFILE_YAML_FORMAT = {
    blankLines: true,
    lineWidth: 1000,
    noCompatMode: true,
    noRefs: true,
    sortKeys: false,
};
/**
 * Parses pnpm-lock.yaml file to `LockFileData` object
 *
 * @param lockFile
 * @returns
 */
function parsePnpmLockFile(lockFile) {
    const data = (0, js_yaml_1.load)(lockFile);
    const { dependencies, devDependencies, packages, specifiers } = data, metadata = tslib_1.__rest(data, ["dependencies", "devDependencies", "packages", "specifiers"]);
    return {
        dependencies: mapPackages(dependencies, devDependencies, specifiers, packages, metadata.lockfileVersion.toString().endsWith('inlineSpecifiers')),
        lockFileMetadata: Object.assign({}, metadata),
        hash: (0, utils_1.hashString)(lockFile),
    };
}
exports.parsePnpmLockFile = parsePnpmLockFile;
function mapPackages(dependencies, devDependencies, specifiers, packages, inlineSpecifiers) {
    const mappedPackages = {};
    Object.entries(packages).forEach(([key, value]) => {
        // construct packageMeta object
        const meta = mapMetaInformation({ dependencies, devDependencies, specifiers }, inlineSpecifiers, key, value);
        // create new key
        const version = key.split('/').pop().split('_')[0];
        const packageName = key.slice(1, key.lastIndexOf('/'));
        const newKey = `${packageName}@${version}`;
        if (!mappedPackages[packageName]) {
            mappedPackages[packageName] = {};
        }
        if (mappedPackages[packageName][newKey]) {
            mappedPackages[packageName][newKey].packageMeta.push(meta);
        }
        else {
            const { dev, optional, peer } = value, rest = tslib_1.__rest(value, ["dev", "optional", "peer"]);
            mappedPackages[packageName][newKey] = Object.assign(Object.assign({}, rest), { version, packageMeta: [meta] });
        }
    });
    Object.keys(mappedPackages).forEach((packageName) => {
        const versions = mappedPackages[packageName];
        const versionKeys = Object.keys(versions);
        if (versionKeys.length === 1) {
            versions[versionKeys[0]].rootVersion = true;
        }
        else {
            const rootVersionKey = versionKeys.find((v) => (0, utils_1.isRootVersion)(packageName, versions[v].version));
            // this should never happen, but just in case
            if (rootVersionKey) {
                versions[rootVersionKey].rootVersion = true;
            }
        }
    });
    return mappedPackages;
}
// maps packageMeta based on dependencies, devDependencies and (inline) specifiers
function mapMetaInformation({ dependencies, devDependencies, specifiers, }, hasInlineSpefiers, key, _a) {
    var { dev, optional, peer } = _a, dependencyDetails = tslib_1.__rest(_a, ["dev", "optional", "peer"]);
    const matchingVersion = key.split('/').pop();
    const packageName = key.slice(1, key.lastIndexOf('/'));
    const isDependency = isVersionMatch(dependencies === null || dependencies === void 0 ? void 0 : dependencies[packageName], matchingVersion, hasInlineSpefiers);
    const isDevDependency = isVersionMatch(devDependencies === null || devDependencies === void 0 ? void 0 : devDependencies[packageName], matchingVersion, hasInlineSpefiers);
    let specifier;
    if (isDependency || isDevDependency) {
        if (hasInlineSpefiers) {
            specifier =
                getSpecifier(dependencies === null || dependencies === void 0 ? void 0 : dependencies[packageName]) ||
                    getSpecifier(devDependencies === null || devDependencies === void 0 ? void 0 : devDependencies[packageName]);
        }
        else {
            if (isDependency || isDevDependency) {
                specifier = specifiers[packageName];
            }
        }
    }
    return Object.assign(Object.assign(Object.assign(Object.assign({ key,
        isDependency,
        isDevDependency,
        specifier }, (dev && { dev })), (optional && { optional })), (peer && { peer })), { dependencyDetails: Object.assign(Object.assign(Object.assign(Object.assign({}, (dependencyDetails.dependencies && {
            dependencies: dependencyDetails.dependencies,
        })), (dependencyDetails.peerDependencies && {
            peerDependencies: dependencyDetails.peerDependencies,
        })), (dependencyDetails.optionalDependencies && {
            optionalDependencies: dependencyDetails.optionalDependencies,
        })), (dependencyDetails.transitivePeerDependencies && {
            transitivePeerDependencies: dependencyDetails.transitivePeerDependencies,
        })) });
}
// version match for dependencies w/ or w/o inline specifier
function isVersionMatch(versionInfo, matchingVersion, hasInlineSpefiers) {
    if (!versionInfo) {
        return false;
    }
    if (!hasInlineSpefiers) {
        return versionInfo === matchingVersion;
    }
    return (versionInfo.version === matchingVersion);
}
function getSpecifier(versionInfo) {
    return (versionInfo && versionInfo.specifier);
}
/**
 * Generates pnpm-lock.yml file from `LockFileData` object
 *
 * @param lockFile
 * @returns
 */
function stringifyPnpmLockFile(lockFileData) {
    const pnpmLockFile = unmapLockFile(lockFileData);
    return (0, js_yaml_1.dump)(pnpmLockFile, LOCKFILE_YAML_FORMAT);
}
exports.stringifyPnpmLockFile = stringifyPnpmLockFile;
// revert lock file to it's original state
function unmapLockFile(lockFileData) {
    const devDependencies = {};
    const dependencies = {};
    const packages = {};
    const specifiers = {};
    const inlineSpecifiers = lockFileData.lockFileMetadata.lockfileVersion
        .toString()
        .endsWith('inlineSpecifiers');
    const packageNames = Object.keys(lockFileData.dependencies);
    for (let i = 0; i < packageNames.length; i++) {
        const packageName = packageNames[i];
        const versions = Object.values(lockFileData.dependencies[packageName]);
        versions.forEach((_a) => {
            var { packageMeta, version: _, rootVersion } = _a, rest = tslib_1.__rest(_a, ["packageMeta", "version", "rootVersion"]);
            packageMeta.forEach(({ key, specifier, isDependency, isDevDependency, dependencyDetails, dev, optional, peer, }) => {
                let version;
                if (inlineSpecifiers) {
                    version = {
                        specifier,
                        version: key.slice(key.lastIndexOf('/') + 1),
                    };
                }
                else {
                    version = key.slice(key.lastIndexOf('/') + 1);
                    if (specifier) {
                        specifiers[packageName] = specifier;
                    }
                }
                if (isDependency) {
                    dependencies[packageName] = version;
                }
                if (isDevDependency) {
                    devDependencies[packageName] = version;
                }
                packages[key] = Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({}, rest), { dev: !!dev }), (optional && { optional })), (peer && { peer })), dependencyDetails);
            });
        });
    }
    const _a = lockFileData.lockFileMetadata, { time } = _a, lockFileMetatada = tslib_1.__rest(_a, ["time"]);
    return Object.assign(Object.assign({}, lockFileMetatada), { specifiers: (0, utils_1.sortObject)(specifiers), dependencies: (0, utils_1.sortObject)(dependencies), devDependencies: (0, utils_1.sortObject)(devDependencies), packages: (0, utils_1.sortObject)(packages), time });
}
/**
 * Returns matching version of the dependency
 */
function transitiveDependencyPnpmLookup({ packageName, versions, version, }) {
    const combinedDependency = Object.values(versions).find((v) => v.packageMeta.some((m) => m.key === `/${packageName}/${version}`));
    if (combinedDependency) {
        return combinedDependency;
    }
    // pnpm's dependencies always point to the exact version so this block is only for insurrance
    return Object.values(versions).find((v) => (0, semver_1.satisfies)(v.version, version));
}
exports.transitiveDependencyPnpmLookup = transitiveDependencyPnpmLookup;
/**
 * Prunes the lock file data based on the list of packages and their transitive dependencies
 *
 * @param lockFileData
 * @returns
 */
function prunePnpmLockFile(lockFileData, packages, projectName) {
    const dependencies = pruneDependencies(lockFileData.dependencies, packages, projectName);
    const prunedLockFileData = {
        lockFileMetadata: pruneMetadata(lockFileData.lockFileMetadata, dependencies),
        dependencies,
        hash: (0, utils_1.generatePrunnedHash)(lockFileData.hash, packages, projectName),
    };
    return prunedLockFileData;
}
exports.prunePnpmLockFile = prunePnpmLockFile;
// iterate over packages to collect the affected tree of dependencies
function pruneDependencies(dependencies, packages, projectName) {
    const result = {};
    packages.forEach((packageName) => {
        if (dependencies[packageName]) {
            const [key, _a] = Object.entries(dependencies[packageName]).find(([_, v]) => v.rootVersion), { packageMeta } = _a, value = tslib_1.__rest(_a, ["packageMeta"]);
            result[packageName] = result[packageName] || {};
            const metaKey = `/${packageName}/${value.version}`;
            const meta = packageMeta.find((m) => m.key.startsWith(metaKey));
            result[packageName][key] = Object.assign(value, {
                packageMeta: [
                    {
                        isDependency: true,
                        key: meta.key,
                        specifier: value.version,
                        dependencyDetails: meta.dependencyDetails,
                    },
                ],
            });
            pruneTransitiveDependencies([packageName], dependencies, result, result[packageName][key]);
        }
        else {
            console.warn(`Could not find ${packageName} in the lock file. Skipping...`);
        }
    });
    return result;
}
function pruneMetadata(lockFileMetadata, prunedDependencies) {
    // These should be removed from the lock file metadata since we don't have them in the package.json
    // overrides, patchedDependencies, neverBuiltDependencies, onlyBuiltDependencies, packageExtensionsChecksum
    return Object.assign({ lockfileVersion: lockFileMetadata.lockfileVersion }, (lockFileMetadata.time && {
        time: pruneTime(lockFileMetadata.time, prunedDependencies),
    }));
}
function pruneTime(time, prunedDependencies) {
    const result = {};
    Object.entries(time).forEach(([key, value]) => {
        const packageName = key.slice(1, key.lastIndexOf('/'));
        const version = key.slice(key.lastIndexOf('/'));
        if (prunedDependencies[packageName] &&
            prunedDependencies[packageName][`${packageName}@${version}`]) {
            result[key] = value;
        }
    });
    return result;
}
// find all transitive dependencies of already pruned packages
// and adds them to the collection
// recursively prune their dependencies
function pruneTransitiveDependencies(parentPackages, dependencies, prunedDeps, parent) {
    Object.entries(Object.assign(Object.assign({}, parent.dependencies), parent.optionalDependencies)).forEach(([packageName, version]) => {
        var _a;
        const versions = dependencies[packageName];
        if (versions) {
            const dependency = transitiveDependencyPnpmLookup({
                packageName,
                parentPackages,
                versions,
                version,
            });
            if (dependency) {
                if (!prunedDeps[packageName]) {
                    prunedDeps[packageName] = {};
                }
                const { packageMeta, rootVersion } = dependency, value = tslib_1.__rest(dependency, ["packageMeta", "rootVersion"]);
                const key = `${packageName}@${value.version}`;
                const meta = findPackageMeta(packageMeta, parentPackages);
                if (prunedDeps[packageName][key]) {
                    // TODO not sure if this is important?
                }
                else {
                    const packageMeta = {
                        key: meta.key,
                        dependencyDetails: meta.dependencyDetails,
                    };
                    prunedDeps[packageName][key] = Object.assign(value, {
                        rootVersion: false,
                        packageMeta: [packageMeta],
                    });
                    if ((_a = parent.optionalDependencies) === null || _a === void 0 ? void 0 : _a[packageName]) {
                        packageMeta.optional = true;
                    }
                    pruneTransitiveDependencies([...parentPackages, packageName], dependencies, prunedDeps, prunedDeps[packageName][key]);
                }
            }
        }
    });
}
function findPackageMeta(packageMeta, parentPackages) {
    const matchPackageVersionModifier = (version) => (packageName) => {
        const normalizedName = packageName.split('/').join('+');
        if (version.includes(`_${normalizedName}@`)) {
            return true;
        }
    };
    const nestedDependency = packageMeta.find((m) => {
        const version = m.key.split('/').pop();
        // it's modified by a single dependency
        // e.g. /@org/my-package/1.0.0_@babel+core@1.2.3
        return (version.includes('_') &&
            parentPackages.some(matchPackageVersionModifier(version)));
    }) ||
        // it's not modified by a single dependency but a mix of them
        // e.g. /@org/my-package/1.0.0_asfgasgasgasg
        packageMeta.find((m) => m.key.split('/').pop().includes('_'));
    if (nestedDependency) {
        return nestedDependency;
    }
    // otherwise it's a root dependency
    return packageMeta.find((m) => !m.key.split('/').pop().includes('_'));
}
